### PR TITLE
Add periodic constraint

### PR DESCRIPTION
**Pull request by HydrogenSulfate**
### PR types
New features

### PR changes
APIs

### Describe
1. 完善`geometry`模块的周期数据点生成函数`periodic_point`
2. 添加周期边界条件 `PeriodicConstraint`
3. 添加3个周期边界条件专用损失函数 `PeriodicL1Loss`，`PeriodicL2Loss`，`PeriodicMSELoss`
4. 支持学习率为普通浮点数的情况，此时lr_scheduler为None，因此所有涉及获取学习率的代码改为`optimizer.get_lr`而不是`lr_scheduler.get_lr`，同时在调节学习率时增加了对lr_scheduler为None的判断
5. 更改所有Constraint中，for循环中使用的变量名`label_expr`为`expr`，防止与入参`label_expr`同名而引起潜在的问题
6. 添加config.py文件，避免在solver.py中import报错

周期边界约束数据构造和损失计算逻辑：
指定batchsize=N（必须是正偶数）时，以一个batch数据为例，会生成N/2个原始点，根据原始点生成N/2个周期点，将原始点和周期点在axis=0上叠加作为真正的输入数据；同理在Loss计算时，使用输出的前半部分（N/2个原始点的对应输出）和后半部分（N/2个周期点的对应输出）计算损失值。由于数据构造要求原始点、周期点严格相间排列，因此需要固定shuffle=False且N为正偶数，避免dataloader根据随机index加载数据而引起错误的周期约束

### CODE REVIEW

**Review by cxxly — APPROVED** [review]
LGTM

**Review by cxxly — APPROVED** [review]
LGTM

**Comment by pecanjk** [comment]
请问对于
Case 1:`u(0,t)=u(1,t)`这种PeriodicBC问题，在PeriodicConstraint里的label_dict这个字段如何设置呢？我的理解是应该可以不用提供label_dict吧？我看代码中`[129]# set all label's to zero for dummy data.`了。
Case 2 `u(0,t)=u(1,t)=f(x)`这种PeriodicBC问题，有周期性约束并且等于同一个值，在PeriodicConstraint里的label_dict这个字段又如何设置呢？

loss是只能使用专用的`PeriodicL1Loss，PeriodicL2Loss，PeriodicMSELoss`么？

**Comment by HydrogenSulfate** [comment]
> PeriodicConstraint

1. 不用提供
2. 你说的这种情况建，周期边界条件比较典型，建议直接对输入做周期化，去掉`u(0,t)=u(1,t)`这条等式，然后剩下的`u(1,t)=f(x)`用普通BC实现即可
image
